Problems with fetch_add for T*

When I try to compile the tests, I get the following error:
```
[ 50%] Building CXX object CMakeFiles/avakar_atomic_ref_test.dir/test/test.cpp.o
In file included from .../atomic_ref/test/test.cpp:1:
.../atomic_ref/include/avakar/atomic_ref.h: In instantiation of ‘T* avakar::_atomic_ref<T*>::fetch_add(avakar::_atomic_ref<T*>::difference_type, std::memory_order) const [with T = long unsigned int; avakar::_atomic_ref<T*>::value_type = long unsigned int*; avakar::_atomic_ref<T*>::difference_type = long int; std::memory_order = std::memory_order]’:
.../atomic_ref/test/test.cpp:14:29:   required from here
.../atomic_ref/include/avakar/atomic_ref.h:217:40: error: no matching function for call to ‘fetch_add(long unsigned int*&, avakar::_atomic_ref<long unsigned int*, void>::difference_type&, std::memory_order&)’
  217 |   return _avakar::atomic_ref::fetch_add(_obj, arg, order);
      |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
In file included from .../atomic_ref/include/avakar/atomic_ref.h:12,
                 from .../atomic_ref/test/test.cpp:1:
.../atomic_ref/include/avakar/../../src/atomic_ref.gcc.h:59:3: note: candidate: ‘template<class T> T _avakar::atomic_ref::fetch_add(T&, T, std::memory_order)’
   59 | T fetch_add(T & obj, T arg, std::memory_order order) noexcept
      |   ^~~~~~~~~
.../atomic_ref/include/avakar/../../src/atomic_ref.gcc.h:59:3: note:   template argument deduction/substitution failed:
In file included from .../atomic_ref/test/test.cpp:1:
.../atomic_ref/include/avakar/atomic_ref.h:217:40: note:   deduced conflicting types for parameter ‘T’ (‘long unsigned int*’ and ‘long int’)
  217 |   return _avakar::atomic_ref::fetch_add(_obj, arg, order);
```

As far as I can tell, the complaint is justified. When using `atomic_ref<T*>::fetch_add()`,  there is no suitable specialization in `atomic_ref.gcc.h`: `fetch_add` gets `T = long unsigend*`, so `obj` is `long unsigned *&`, which is fine, but the arg with `long unsigned*` doesn't make sense, since it's just an integer `difference_type` in the caller.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,5 +1,6 @@
 cmake_minimum_required(VERSION 3.12)
 project(avakar_atomic_ref)
+include(FetchContent)
 
 add_library(avakar_atomic_ref INTERFACE)
 target_include_directories(avakar_atomic_ref INTERFACE include)
@@ -9,7 +10,24 @@ add_library(avakar::atomic_ref ALIAS avakar_atomic_ref)
 
 include(CTest)
 if (BUILD_TESTING)
-	add_executable(avakar_atomic_ref_test test/test.cpp)
-	target_link_libraries(avakar_atomic_ref_test avakar::atomic_ref)
+	FetchContent_Declare(
+		Catch2
+		GIT_REPOSITORY https://github.com/catchorg/Catch2.git
+		GIT_TAG d10b9bd02e098476670f5eb0527d2c7281476e8a
+		GIT_SHALLOW YES
+		)
+
+	FetchContent_GetProperties(Catch2)
+	if (NOT catch2_POPULATED)
+		FetchContent_Populate(Catch2)
+		add_subdirectory("${catch2_SOURCE_DIR}" "${catch2_BINARY_DIR}")
+	endif()
+
+	add_executable(avakar_atomic_ref_test
+		test/main.cpp
+		test/test.cpp
+		)
+	target_link_libraries(avakar_atomic_ref_test avakar::atomic_ref Catch2::Catch2)
+
 	add_test(NAME avakar::atomic_ref COMMAND avakar_atomic_ref_test)
 endif()

--- a/src/atomic_ref.gcc.h
+++ b/src/atomic_ref.gcc.h
@@ -55,34 +55,46 @@ bool compare_exchange_strong(T & obj, T & expected, T desired, std::memory_order
 	return __atomic_compare_exchange(&obj, &expected, &desired, false, success, failure);
 }
 
-template <typename T, typename U>
-T fetch_add(T & obj, U arg, std::memory_order order) noexcept
+template <typename T>
+T fetch_add(T & obj, T arg, std::memory_order order) noexcept
 {
 	return __atomic_fetch_add(&obj, arg, order);
 }
 
-template <typename T, typename U>
-T fetch_sub(T & obj, U arg, std::memory_order order) noexcept
+template <typename T>
+T fetch_sub(T & obj, T arg, std::memory_order order) noexcept
 {
 	return __atomic_fetch_sub(&obj, arg, order);
 }
 
-template <typename T, typename U>
-T fetch_and(T & obj, U arg, std::memory_order order) noexcept
+template <typename T>
+T fetch_and(T & obj, T arg, std::memory_order order) noexcept
 {
 	return __atomic_fetch_and(&obj, arg, order);
 }
 
-template <typename T, typename U>
-T fetch_or(T & obj, U arg, std::memory_order order) noexcept
+template <typename T>
+T fetch_or(T & obj, T arg, std::memory_order order) noexcept
 {
 	return __atomic_fetch_or(&obj, arg, order);
 }
 
-template <typename T, typename U>
-T fetch_xor(T & obj, U arg, std::memory_order order) noexcept
+template <typename T>
+T fetch_xor(T & obj, T arg, std::memory_order order) noexcept
 {
 	return __atomic_fetch_xor(&obj, arg, order);
+}
+
+template <typename T>
+auto fetch_add(T * & obj, std::ptrdiff_t arg, std::memory_order order) noexcept
+{
+	return __atomic_fetch_add(&obj, arg * sizeof(T), order);
+}
+
+template <typename T>
+auto fetch_sub(T * & obj, std::ptrdiff_t arg, std::memory_order order) noexcept
+{
+	return __atomic_fetch_sub(&obj, arg * sizeof(T), order);
 }
 
 }

--- a/test/main.cpp
+++ b/test/main.cpp
@@ -1,0 +1,2 @@
+#define CATCH_CONFIG_MAIN
+#include <catch2/catch.hpp>

--- a/test/test.cpp
+++ b/test/test.cpp
@@ -1,17 +1,40 @@
 #include <avakar/atomic_ref.h>
-#include <stdint.h>
+#include <catch2/catch.hpp>
+using avakar::atomic_ref;
 
-int main()
+TEST_CASE("Pointer add/sub is correct")
 {
-	using avakar::atomic_ref;
+	int v = 0;
+	atomic_ref<int> a(v);
 
-	uint64_t k = 0;
-	atomic_ref<uint64_t> ak(k);
+	REQUIRE(a.fetch_add(1) == 0);
+	REQUIRE(a.fetch_add(1) == 1);
+	REQUIRE(v == 2);
+	REQUIRE(a.fetch_sub(1) == 2);
+	REQUIRE(a.fetch_sub(1) == 1);
+	REQUIRE(v == 0);
 
-	uint64_t * ptr = &k;
-	atomic_ref<uint64_t *> aptr(ptr);
+	REQUIRE(a.fetch_add(2) == 0);
+	REQUIRE(v == 2);
+	REQUIRE(a.fetch_sub(2) == 2);
+	REQUIRE(v == 0);
+}
 
-	auto ppp = aptr.fetch_add(1);
+TEST_CASE("Pointer arithmetic is correct")
+{
+	int arr[] = { 1, 2, 3 };
+	int * p = arr;
+	atomic_ref<int *> a(p);
 
-	return (int)ak.exchange(1);
+	REQUIRE(a.fetch_add(1) == &arr[0]);
+	REQUIRE(a.fetch_add(1) == &arr[1]);
+	REQUIRE(p == &arr[2]);
+	REQUIRE(a.fetch_sub(1) == &arr[2]);
+	REQUIRE(a.fetch_sub(1) == &arr[1]);
+	REQUIRE(p == &arr[0]);
+
+	REQUIRE(a.fetch_add(2) == &arr[0]);
+	REQUIRE(p == &arr[2]);
+	REQUIRE(a.fetch_sub(2) == &arr[2]);
+	REQUIRE(p == &arr[0]);
 }

